core/src/main/java/com/linecorp/armeria/server/websocket/DefaultWebSocketService.java

```diff
@@ -0,0 +1,396 @@
+                     .filter(sub -> SUB_PROTOCOL_WILDCARD.equals(sub) ||
+                                    this.subprotocols.contains(sub))
+                     .findFirst().ifPresent(selectedSubprotocol -> responseHeadersBuilder.add(
+                             HttpHeaderNames.SEC_WEBSOCKET_PROTOCOL, selectedSubprotocol));
+    }
+
+    /**
+     * Handles the HTTP/2 web socket handshake described in
+     * <a href="https://datatracker.ietf.org/doc/html/rfc8441">Bootstrapping WebSockets with HTTP/2</a>.
+     * These are examples of a request and the corresponding response:
+     *
+     * <p>Request:
+     * <pre>
+     * HEADERS + END_HEADERS
+     * :method = CONNECT
+     * :protocol = websocket
+     * :scheme = https
+     * :path = /chat
+     * :authority = server.example.com
+     * sec-websocket-protocol = chat, superchat
+     * sec-websocket-extensions = permessage-deflate
+     * sec-websocket-version = 13
+     * origin = http://www.example.com
+     * </pre>
+     *
+     * <p>Response:
+     * <pre>
+     * HEADERS + END_HEADERS
+     * :status = 200
+     * sec-websocket-protocol = chat
+     * </pre>
+     */
+    private WebSocketUpgradeResult upgradeHttp2(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        if (!ctx.sessionProtocol().isExplicitHttp2()) {
+            final HttpResponse fallbackResponse;
+            if (fallbackService != null) {
+                fallbackResponse = fallbackService.serve(ctx, req);
+            } else {
+                fallbackResponse = HttpResponse.of(HttpStatus.METHOD_NOT_ALLOWED);
+            }
+            return WebSocketUpgradeResult.ofFailure(fallbackResponse);
+        }
+        final RequestHeaders headers = req.headers();
+        if (!isHttp2WebSocketUpgradeRequest(headers)) {
+            logger.trace("RequestHeaders does not contain headers for WebSocket upgrade. headers: {}", headers);
+            final HttpResponse fallbackResponse;
+            if (fallbackService != null) {
+                fallbackResponse = fallbackService.serve(ctx, req);
+            } else {
+                fallbackResponse = HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT_UTF_8,
+                                                   "The upgrade header must contain:\n" +
+                                                   "  :protocol = websocket");
+            }
+            return WebSocketUpgradeResult.ofFailure(fallbackResponse);
+        }
+
+        HttpResponse invalidResponse = checkOrigin(ctx, headers);
+        if (invalidResponse != null) {
+            return WebSocketUpgradeResult.ofFailure(invalidResponse);
+        }
+
+        invalidResponse = checkVersion(headers);
+        if (invalidResponse != null) {
+            return WebSocketUpgradeResult.ofFailure(invalidResponse);
+        }
+
+        return WebSocketUpgradeResult.ofSuccess();
+    }
+
+    private HttpResponse maybeFallbackResponse(ServiceRequestContext ctx, HttpRequest req,
```

Comment:
nit: This method isn't used.

Reply:
Nice catch! It was superseded by `failOrFallback`.